src/vehicle_gui.cpp

```diff
@@ -329,6 +329,82 @@ static bool CargoFilter(const GUIVehicleGroup *vehgroup, const CargoID cid)
 	return false;
 }
 
+/**
+ * Test if cargo icon overlays should be drawn.
+ * @returns true iff cargo icon overlays should be drawn.
+ */
+bool ShowCargoIconOverlay()
+{
+	return _shift_pressed && _ctrl_pressed;
+}
+
+/**
+ * Add a cargo icon to the list of overlays.
+ * @param overlays List of overlays.
+ * @param x Horizontal position.
+ * @param width Width available.
+ * @param v Vehicle to add.
+ */
+void AddCargoIconOverlay(std::vector<CargoIconOverlay> &overlays, int x, int width, const Vehicle *v)
+{
+	bool rtl = _current_text_dir == TD_RTL;
+	if (!v->IsArticulatedPart() || v->cargo_type != v->Previous()->cargo_type) {
+		/* Add new overlay slot. */
+		overlays.emplace_back(rtl ? x - width : x, rtl ? x : x + width, v->cargo_type, v->cargo_cap);
+	} else {
+		/* This is an articulated part with the same cargo type, adjust left or right of last overlay slot. */
+		if (rtl) {
+			overlays.back().left -= width;
+		} else {
+			overlays.back().right += width;
+		}
+		overlays.back().cargo_cap += v->cargo_cap;
```

Comment:
If the vehicle's capacity is 0, the overlay doesn't get drawn. Why add this unused overlay to the vector at all? 

If we need to track it (I assume it's related to articulated vehicles), why store the actual cargo capacity over a bool to mark the segment as not having an icon? (I might be over-optimizing here.)

Reply:
It is indeed added so that if an articulated part is present and is the same cargo type we can extend the slot from the first part of the articulated vehicle.

While it is possible to test if a part has an articulated part, and then add or not add accordingly, and then maybe remove an entry if it turns out the articulated vehicle has no cargo... the code becomes quite a bit more complex.

The overlays vector is marked static to avoid so that the reserved memory is reused and does not need fresh allocations every time.

The capacity could just be stored as a bool, but simply adding capacity saves me from caring about conditionals until it comes to drawing. Potentially less branch prediction happening, but not really my primary thought...